SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -0,0 +1,338 @@
+        OWSAssert(viewTransaction != nil);
+        job = [viewTransaction firstObjectInGroup:OWSMessageProcessingJobFinderExtensionGroup];
+    }];
+
+    return job;
+}
+
+- (void)addJobForEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
+{
+    [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+        [[[OWSMessageProcessingJob alloc] initWithEnvelope:envelope] saveWithTransaction:transaction];
+    }];
+}
+
+- (void)removeJobWithId:(NSString *)uniqueId
+{
+    [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+        [transaction removeObjectForKey:uniqueId inCollection:[OWSMessageProcessingJob collection]];
+    }];
+}
+
++ (YapDatabaseView *)databaseExension
+{
+
+    YapDatabaseViewSorting *sorting =
+        [YapDatabaseViewSorting withObjectBlock:^NSComparisonResult(YapDatabaseReadTransaction *transaction,
+            NSString *group,
+            NSString *collection1,
+            NSString *key1,
+            id object1,
+            NSString *collection2,
+            NSString *key2,
+            id object2) {
+
+            if (![object1 isKindOfClass:[OWSMessageProcessingJob class]]) {
+                OWSFail(@"Unexpected object: %@ in collection: %@", object1, collection1) return NSOrderedSame;
+            }
+            OWSMessageProcessingJob *job1 = (OWSMessageProcessingJob *)object1;
+
+            if (![object2 isKindOfClass:[OWSMessageProcessingJob class]]) {
+                OWSFail(@"Unexpected object: %@ in collection: %@", object2, collection2) return NSOrderedSame;
+            }
+            OWSMessageProcessingJob *job2 = (OWSMessageProcessingJob *)object2;
+
+            return [job1.createdAt compare:job2.createdAt];
+        }];
+
+    YapDatabaseViewGrouping *grouping =
+        [YapDatabaseViewGrouping withObjectBlock:^NSString *_Nullable(YapDatabaseReadTransaction *_Nonnull transaction,
+            NSString *_Nonnull collection,
+            NSString *_Nonnull key,
+            id _Nonnull object) {
+            if (![object isKindOfClass:[OWSMessageProcessingJob class]]) {
+                OWSFail(@"Unexpected object: %@ in collection: %@", object, collection) return nil;
+            }
+
+            // Arbitrary string - all in the same group. We're only using the view for sorting.
+            return OWSMessageProcessingJobFinderExtensionGroup;
+        }];
+
+    YapDatabaseViewOptions *options = [YapDatabaseViewOptions new];
+    options.allowedCollections =
+        [[YapWhitelistBlacklist alloc] initWithWhitelist:[NSSet setWithObject:[OWSMessageProcessingJob collection]]];
+
+    return [[YapDatabaseView alloc] initWithGrouping:grouping sorting:sorting versionTag:@"1" options:options];
+}
+
+
++ (void)syncRegisterDatabaseExtension:(YapDatabase *)database
+{
+    YapDatabaseView *existingView = [database registeredExtension:OWSMessageProcessingJobFinderExtensionName];
+    if (existingView) {
```

Comment:
Don't we expect to be registered exactly once?  Can we assert or fail if its already registered?  That would indicate, for example, that perhaps two views have the same name or something.